crates/rattler_conda_types/src/conda_lock/mod.rs

```diff
@@ -0,0 +1,222 @@
+//! Most names were kept the same as in the models file. So you can refer to those exactly.
+//! However, some types were added to enforce a bit more type safety.
+
+use crate::conda_lock::PackageHashes::{Md5, Md5Sha256, Sha256};
+use crate::Platform;
+use serde::de::Error;
+use serde::{Deserialize, Deserializer, Serialize};
+use std::collections::HashMap;
+use url::Url;
+
+/// Default version for the conda-lock file format
+const fn default_version() -> u32 {
+    1
+}
+
+/// Represents the conda-lock file
+/// Contains the metadata regarding the lock files
+/// also the locked packages
+#[derive(Serialize, Deserialize)]
+pub struct CondaLock {
+    /// Metadata for the lock file
+    pub metadata: LockMeta,
+    /// Locked packages
+    pub package: Vec<LockedDependency>,
+    #[serde(default = "default_version")]
+    pub version: u32,
+}
+
+#[derive(Serialize, Deserialize)]
+/// Metadata for the [`CondaLock`] file
+pub struct LockMeta {
+    /// Hash of dependencies for each target platform
+    pub content_hash: HashMap<Platform, String>,
+    /// Channels used to resolve dependencies
+    pub channels: Vec<Channel>,
+    /// The platforms this lock file supports
+    pub platforms: Vec<Platform>,
+    /// Paths to source files, relative to the parent directory of the lockfile
+    pub sources: Vec<String>,
+    /// Metadata dealing with the time lockfile was created
+    pub time_metadata: Option<TimeMeta>,
+    /// Metadata dealing with the git repo the lockfile was created in and the user that created it
+    pub git_metadata: Option<GitMeta>,
+    /// Metadata dealing with the input files used to create the lockfile
+    pub inputs_metadata: Option<HashMap<String, PackageHashes>>,
+    /// Custom metadata provided by the user to be added to the lockfile
+    pub custom_metadata: Option<HashMap<String, String>>,
+}
+
+/// Stores information about when the lockfile was generated
+#[derive(Serialize, Deserialize)]
+pub struct TimeMeta {
+    /// Time stamp of lock-file creation format
+    // TODO: I think this is UTC time, change this later, conda-lock is not really using this now
+    pub created_at: String,
+}
+
+/// Stores information about the git repo the lockfile is being generated in (if applicable) and
+/// the git user generating the file.
+#[derive(Serialize, Deserialize)]
+pub struct GitMeta {
+    /// Git user.name field of global config
+    pub git_user_name: String,
+    /// Git user.email field of global config
+    pub git_user_email: String,
+    /// sha256 hash of the most recent git commit that modified one of the input files for this lockfile
+    pub git_sha: String,
+}
+
+#[derive(Serialize, Deserialize)]
+/// Represents whether this is a dependency managed by pip or conda
+pub enum Manager {
+    #[serde(rename = "conda")]
+    Conda,
+    #[serde(rename = "pip")]
+    Pip,
+}
+
+#[derive(Serialize, Deserialize, Eq, PartialEq, Hash)]
```

Comment:
Why did you use new types here? Especially for package name String looks pretty logical. Maybe you can at least implement Display for it?

There is now also no way to access the string name because the string is not `pub`.

Reply:
Removed `PackageName` added Display for `VersionConstraint`